e2e/__tests__/jest.config.ts.test.ts

```diff
@@ -0,0 +1,73 @@
+/**
```

Comment:
A few tests it would be good to have:

* `it does not type check the config` (I'd have a config that does something like cast a number to a string)
  * bonus points if you do that with a config that jest fails on, to also implicitly show that validation should still take place - I don't think that is worth it's own test, since it should always take place, but can be nice to have it covered even a little
* what happens if you have a tsconfig, or a nest of them?
* testing what happens if you have the target set to different things, mainly thinking of `commonjs` vs `es2015` output.

Those last two might be really hard to test - I don't think they're super critical for this to land, because `jest.config.ts` with some restrictions (i.e you have to do `export default`) is better than no support at all.

What I'm conscious of is that when doing webpack configs typically your tsconfig.json is set to `es2015`, which can cause `ts-node` to be upset since it can't actually do that target.

In saying that, I've not actually done a lot of configuring recently and don't have repos on hand that I can do some quick testing myself, so it could all be fine. `ts-node` now supports being configured via its own property in `tsconfig.json`, which makes things more flexible anyway.

But that first for the typechecking I think is a must.

Reply:
@G-Rath I understood the last 2 points and I will try to do some tests although don't think I have time to go in detail with the implementation.

About `it does not type check the config`, I don't quite understand, because in my mind the types should be checked on parsing.

Reply:
In that case, we should have a test that does the opposite: `it does type check the config`.

My thinking is that `jest` itself doesn't need to do type checking because we can use `typescript` for that - jest will still validate the config to ensure it's valid enough to be used, which will give nicer error messages (i.e jest will say "this property should be a path" when typescript would just say "type number is not assignable to type string").

(In addition to nicer error messages, it'd also be slightly faster since we'll be transpiling only, but not by a lot since it's a one-time cost at start up).

I think the pros vs cons on typechecking vs not are pretty evenly matched, so it's fine to stick with type-checking 🙂